### PR TITLE
Adding app inbound auth migration client

### PR DESCRIPTION
**Pull request by AyoobAboosalih**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import org.wso2.carbon.is.migration.service.v530.util.JDBCPersistenceUtil;
+import org.wso2.carbon.is.migration.util.Schema;
+
+import java.sql.*;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.TimeZone;
+
+/**
+ * Migration implementation for migrating recovery data.
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:35]: Update the javadoc description.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                appids.put(appid, count);
+            }
+            return appids;
+        } catch (SQLException e) {
+            throw new MigrationClientException("An error occurred while reading auth data.", e);
+        }
+    }
+
+    // Function that contains the logic to delete duplicate data entries
+    private void deleteDuplicateData(HashMap<Integer, Integer>  appids) throws MigrationClientException {
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:88]: `deleteStaleInboundAuthEntries` would be a better name for this method.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    @Override
+    public void migrate() throws MigrationClientException {
+
+        log.info("Started migrating app inbound authentication data.");
+        HashMap<Integer, Integer> appids = getAppIDCountFromSPInboundAuthTable();
+        deleteDuplicateData(appids);
+        log.info("Inbound authentication data migration complete.");
+    }
+
+    // Function to retrieve app ID's along with their corresponding count.
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:67]: Lets follow the javadoc comment format.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        } catch (SQLException e) {
+            throw new MigrationClientException("An error occurred while reading auth data.", e);
+        }
+    }
+
+    // Function that contains the logic to delete duplicate data entries
+    private void deleteDuplicateData(HashMap<Integer, Integer>  appids) throws MigrationClientException {
+
+        if (!appids.isEmpty()) {
+            log.info("Removing duplicate service provider entries.");
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:91]: Change `duplicate` to `stale`

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            return appids;
+        } catch (SQLException e) {
+            throw new MigrationClientException("An error occurred while reading auth data.", e);
+        }
+    }
+
+    // Function that contains the logic to delete duplicate data entries
+    private void deleteDuplicateData(HashMap<Integer, Integer>  appids) throws MigrationClientException {
+
+        if (!appids.isEmpty()) {
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:90]: If appIds IS empty, this method needs to return immediately with a log saying that there are no stale entries.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            log.info("Removing duplicate service provider entries.");
+        }
+        try (Connection connection = getDataSource(Schema.IDENTITY.getName()).getConnection()) {
+            boolean autoCommitStatus = connection.getAutoCommit();
+            connection.setAutoCommit(false);
+            for (Integer code : appids.keySet()) {
+                Integer count = appids.get(code);
+                if (count == 3){
+                    try (PreparedStatement preparedStatement = connection.prepareStatement(DELETE_MULTIPLE_AUTH_ENTRIES)) {
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:99]: Line is longer than 120 characters.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                }
+                else if (count == 2) {
+                    try (PreparedStatement retrieveAuthData = connection.prepareStatement(GET_AUTH_DATA)) {
+                        retrieveAuthData.setInt(1,code);
+                        retrieveAuthData.setString(2,PASSIVESTS);
+                        ResultSet resultSet = retrieveAuthData.executeQuery();
+                        if (resultSet.next()) {
+                            String propName = resultSet.getString("PROP_NAME");
+                            if ("passiveSTSWReply".equalsIgnoreCase(propName)) {
+                                try (PreparedStatement preparedStatement =
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:118]: We can move the try catch block outside of the propName check if block and use the if block to set the parameters according to the condition and execute the statement. The `preparedStatement` can be recycled that way with a single catch clause for both conditions.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+ * Migration implementation for migrating recovery data.
+ */
+public class AppInboundAuthMigrator extends Migrator {
+
+    private static final Logger log = LoggerFactory.getLogger(AppInboundAuthMigrator.class);
+    private static final String PASSIVESTS = "passivests";
+    private static final String OPENID = "openid";
+    private static final String GET_APP_ID_COUNT =
+            "SELECT APP_ID, COUNT(APP_ID) as count FROM SP_INBOUND_AUTH GROUP BY APP_ID HAVING COUNT(APP_ID) >= 2";
+    private static final String ERROR_DELETING_AUTH_DATA = "An error occurred while deleting auth data.";
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java:44]: Lets change this to `An error occurred while deleting service provider inbound auth data.`

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 1 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 2 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 3 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 4 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 5 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 6 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 7 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java (hunk 8 above)
Comment (AyoobAboosalih) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v600/migrator/AppInboundAuthMigrator.java]: adjusted with https://github.com/wso2-extensions/identity-migration-resources/pull/262/commits/5d099513c9c31cd2f25bbe3e9ed7a784a85b5f38

**Review by ashendes — APPROVED** [review]

**Pull request merged**